MySQL CI: Incorrect integer value error when running TestRunSQLProgram and TestExecuteXGBoostClassifier

Based on https://github.com/sql-machine-learning/sqlflow/pull/2970 .

```
2020-09-30T01:48:08.3417713Z [0m[0m=== RUN   TestRunSQLProgram
2020-09-30T01:48:18.3652566Z [0m[91m--- FAIL: TestRunSQLProgram (10.03s)
2020-09-30T01:48:18.3653819Z [0m[0m    executor_ir_test.go:143: 
2020-09-30T01:48:18.3654715Z [0m[0m        	Error Trace:	executor_ir_test.go:143
2020-09-30T01:48:18.3655800Z [0m[0m        	            				executor_ir_test.go:123
2020-09-30T01:48:18.3656778Z [0m[0m        	Error:      	Should be true
2020-09-30T01:48:18.3657643Z [0m[0m        	Test:       	TestRunSQLProgram
2020-09-30T01:48:18.3658291Z [0m[0m        	Messages:   	runSQLProgram error: exit status 1
2020-09-30T01:48:18.3659766Z [0m[0m        	            	[09:48:17] SparsePage::Writer Finished writing to /tmp/ordj88o4/predict.txt_0.cache.row.page
2020-09-30T01:48:18.3660654Z [0m[0m        	            	Traceback (most recent call last):
2020-09-30T01:48:18.3661564Z [0m[0m        	            	  File "<stdin>", line 13, in <module>
2020-09-30T01:48:18.3662529Z [0m[0m        	            	  File "/home/sqlflow/actions-runner/_work/sqlflow/sqlflow/python/runtime/local/submitter.py", line 116, in submit_local_pred
2020-09-30T01:48:18.3663430Z [0m[0m        	            	    model=model)
2020-09-30T01:48:18.3664327Z [0m[0m        	            	  File "/home/sqlflow/actions-runner/_work/sqlflow/sqlflow/python/runtime/local/xgboost_submitter/predict.py", line 93, in pred
2020-09-30T01:48:18.3665218Z [0m[0m        	            	    train_label_idx, feature_file_name, conn)
2020-09-30T01:48:18.3666254Z [0m[0m        	            	  File "/home/sqlflow/actions-runner/_work/sqlflow/sqlflow/python/runtime/local/xgboost_submitter/predict.py", line 156, in _store_predict_result
2020-09-30T01:48:18.3667080Z [0m[0m        	            	    line_no += 1
2020-09-30T01:48:18.3668363Z [0m[0m        	            	  File "/usr/lib/python3.6/contextlib.py", line 88, in __exit__
2020-09-30T01:48:18.3669373Z [0m[0m        	            	    next(self.gen)
2020-09-30T01:48:18.3670488Z [0m[0m        	            	  File "/home/sqlflow/actions-runner/_work/sqlflow/sqlflow/python/runtime/db.py", line 299, in buffered_db_writer
2020-09-30T01:48:18.3671720Z [0m[0m        	            	    w.close()
2020-09-30T01:48:18.3673073Z [0m[0m        	            	  File "/home/sqlflow/actions-runner/_work/sqlflow/sqlflow/python/runtime/db_writer/base.py", line 38, in close
2020-09-30T01:48:18.3674376Z [0m[0m        	            	    self.flush()
2020-09-30T01:48:18.3675226Z [0m[0m        	            	  File "/home/sqlflow/actions-runner/_work/sqlflow/sqlflow/python/runtime/db_writer/mysql.py", line 43, in flush
2020-09-30T01:48:18.3676180Z [0m[0m        	            	    cursor.executemany(self.statement, self.rows)
2020-09-30T01:48:18.3677334Z [0m[0m        	            	  File "/home/sqlflow/actions-runner/_work/sqlflow/sqlflow/build/env/lib/python3.6/site-packages/MySQLdb/cursors.py", line 237, in executemany
2020-09-30T01:48:18.3678240Z [0m[0m        	            	    self._get_db().encoding)
2020-09-30T01:48:18.3679241Z [0m[0m        	            	  File "/home/sqlflow/actions-runner/_work/sqlflow/sqlflow/build/env/lib/python3.6/site-packages/MySQLdb/cursors.py", line 264, in _do_execute_many
2020-09-30T01:48:18.3680500Z [0m[0m        	            	    rows += self.execute(sql + postfix)
2020-09-30T01:48:18.3681519Z [0m[0m        	            	  File "/home/sqlflow/actions-runner/_work/sqlflow/sqlflow/build/env/lib/python3.6/site-packages/MySQLdb/cursors.py", line 209, in execute
2020-09-30T01:48:18.3682387Z [0m[0m        	            	    res = self._query(query)
2020-09-30T01:48:18.3683421Z [0m[0m        	            	  File "/home/sqlflow/actions-runner/_work/sqlflow/sqlflow/build/env/lib/python3.6/site-packages/MySQLdb/cursors.py", line 315, in _query
2020-09-30T01:48:18.3684252Z [0m[0m        	            	    db.query(q)
2020-09-30T01:48:18.3685224Z [0m[0m        	            	  File "/home/sqlflow/actions-runner/_work/sqlflow/sqlflow/build/env/lib/python3.6/site-packages/MySQLdb/connections.py", line 226, in query
2020-09-30T01:48:18.3686239Z [0m[0m        	            	    _mysql.connection.query(self, query)
2020-09-30T01:48:18.3687376Z [0m[0m        	            	MySQLdb._exceptions.OperationalError: (1366, "Incorrect integer value: '[0.6262421  0.2188864  0.15487151]' for column `iris`.`predict`.`class` at row 1")
2020-09-30T01:48:18.3688282Z [0m[0m        	            	: [25]	train-merror:0.045455
2020-09-30T01:48:18.3688731Z [0m[0m        	            	
2020-09-30T01:48:18.3689181Z [0m[0m        	            	[26]	train-merror:0.045455
2020-09-30T01:48:18.3689614Z [0m[0m        	            	
2020-09-30T01:48:18.3690079Z [0m[0m        	            	[27]	train-merror:0.045455
2020-09-30T01:48:18.3690569Z [0m[0m        	            	
2020-09-30T01:48:18.3691034Z [0m[0m        	            	[28]	train-merror:0.045455
2020-09-30T01:48:18.3691573Z [0m[0m        	            	
2020-09-30T01:48:18.3692055Z [0m[0m        	            	[29]	train-merror:0.045455
2020-09-30T01:48:18.3692484Z [0m[0m        	            	
2020-09-30T01:48:18.3693457Z [0m[0m        	            	Evaluation result: {'train': {'merror': [0.145455, 0.109091, 0.136364, 0.118182, 0.118182, 0.1, 0.090909, 0.090909, 0.081818, 0.081818, 0.081818, 0.081818, 0.081818, 0.081818, 0.072727, 0.072727, 0.054545, 0.063636, 0.054545, 0.054545, 0.054545, 0.045455, 0.045455, 0.045455, 0.045455, 0.045455, 0.045455, 0.045455, 0.045455, 0.045455]}}
2020-09-30T01:48:18.3694028Z [0m[0m        	            	
2020-09-30T01:48:18.3694556Z [0m[0m        	            	Start predicting XGBoost model...
2020-09-30T01:48:18.3695023Z [0m[0m        	            	
2020-09-30T01:48:18.3696188Z [0m[0m        	            	[09:48:17] SparsePageSource::CreateRowPage Finished writing to /tmp/ordj88o4/predict.txt_0.cache
2020-09-30T01:48:18.3696897Z [0m[0m        	            	
2020-09-30T01:48:18.3697671Z [0m[0m        	            	[09:48:17] 10x2 matrix with 10 entries loaded from /tmp/ordj88o4/predict.txt_0#/tmp/ordj88o4/predict.txt_0.cache
2020-09-30T01:48:18.3698567Z [0m[0m=== RUN   TestExecuteXGBoostClassifier
2020-09-30T01:48:47.0504614Z [0m[91m--- FAIL: TestExecuteXGBoostClassifier (28.69s)
2020-09-30T01:48:47.0509859Z [0m[0m    executor_ir_test.go:154: 
2020-09-30T01:48:47.0515411Z [0m[0m        	Error Trace:	executor_ir_test.go:154
2020-09-30T01:48:47.0516535Z [0m[0m        	            				executor_ir_test.go:150
2020-09-30T01:48:47.0517457Z [0m[0m        	Error:      	Should be true
2020-09-30T01:48:47.0519508Z [0m[0m        	Test:       	TestExecuteXGBoostClassifier
2020-09-30T01:48:47.0521003Z [0m[0m        	Messages:   	runSQLProgram error: exit status 1
2020-09-30T01:48:47.0522575Z [0m[0m        	            	[09:48:27] SparsePage::Writer Finished writing to /tmp/kouvjl02/predict.txt_0.cache.row.page
2020-09-30T01:48:47.0523461Z [0m[0m        	            	Traceback (most recent call last):
2020-09-30T01:48:47.0524162Z [0m[0m        	            	  File "<stdin>", line 13, in <module>
2020-09-30T01:48:47.0525061Z [0m[0m        	            	  File "/home/sqlflow/actions-runner/_work/sqlflow/sqlflow/python/runtime/local/submitter.py", line 116, in submit_local_pred
2020-09-30T01:48:47.0526193Z [0m[0m        	            	    model=model)
2020-09-30T01:48:47.0527065Z [0m[0m        	            	  File "/home/sqlflow/actions-runner/_work/sqlflow/sqlflow/python/runtime/local/xgboost_submitter/predict.py", line 93, in pred
2020-09-30T01:48:47.0527943Z [0m[0m        	            	    train_label_idx, feature_file_name, conn)
2020-09-30T01:48:47.0529485Z [0m[0m        	            	  File "/home/sqlflow/actions-runner/_work/sqlflow/sqlflow/python/runtime/local/xgboost_submitter/predict.py", line 156, in _store_predict_result
2020-09-30T01:48:47.0530910Z [0m[0m        	            	    line_no += 1
2020-09-30T01:48:47.0531978Z [0m[0m        	            	  File "/usr/lib/python3.6/contextlib.py", line 88, in __exit__
2020-09-30T01:48:47.0532627Z [0m[0m        	            	    next(self.gen)
2020-09-30T01:48:47.0533486Z [0m[0m        	            	  File "/home/sqlflow/actions-runner/_work/sqlflow/sqlflow/python/runtime/db.py", line 299, in buffered_db_writer
2020-09-30T01:48:47.0534216Z [0m[0m        	            	    w.close()
2020-09-30T01:48:47.0535001Z [0m[0m        	            	  File "/home/sqlflow/actions-runner/_work/sqlflow/sqlflow/python/runtime/db_writer/base.py", line 38, in close
2020-09-30T01:48:47.0535713Z [0m[0m        	            	    self.flush()
2020-09-30T01:48:47.0536507Z [0m[0m        	            	  File "/home/sqlflow/actions-runner/_work/sqlflow/sqlflow/python/runtime/db_writer/mysql.py", line 43, in flush
2020-09-30T01:48:47.0537414Z [0m[0m        	            	    cursor.executemany(self.statement, self.rows)
2020-09-30T01:48:47.0538718Z [0m[0m        	            	  File "/home/sqlflow/actions-runner/_work/sqlflow/sqlflow/build/env/lib/python3.6/site-packages/MySQLdb/cursors.py", line 237, in executemany
2020-09-30T01:48:47.0539615Z [0m[0m        	            	    self._get_db().encoding)
2020-09-30T01:48:47.0540664Z [0m[0m        	            	  File "/home/sqlflow/actions-runner/_work/sqlflow/sqlflow/build/env/lib/python3.6/site-packages/MySQLdb/cursors.py", line 264, in _do_execute_many
2020-09-30T01:48:47.0541592Z [0m[0m        	            	    rows += self.execute(sql + postfix)
2020-09-30T01:48:47.0542566Z [0m[0m        	            	  File "/home/sqlflow/actions-runner/_work/sqlflow/sqlflow/build/env/lib/python3.6/site-packages/MySQLdb/cursors.py", line 209, in execute
2020-09-30T01:48:47.0543474Z [0m[0m        	            	    res = self._query(query)
2020-09-30T01:48:47.0544403Z [0m[0m        	            	  File "/home/sqlflow/actions-runner/_work/sqlflow/sqlflow/build/env/lib/python3.6/site-packages/MySQLdb/cursors.py", line 315, in _query
2020-09-30T01:48:47.0545205Z [0m[0m        	            	    db.query(q)
2020-09-30T01:48:47.0546105Z [0m[0m        	            	  File "/home/sqlflow/actions-runner/_work/sqlflow/sqlflow/build/env/lib/python3.6/site-packages/MySQLdb/connections.py", line 226, in query
2020-09-30T01:48:47.0547082Z [0m[0m        	            	    _mysql.connection.query(self, query)
2020-09-30T01:48:47.0549468Z [0m[0m        	            	MySQLdb._exceptions.OperationalError: (1366, "Incorrect integer value: '[0.06508882 0.12504072 0.8098704 ]' for column `iris`.`predict`.`class` at row 1")
2020-09-30T01:48:47.0550516Z [0m[0m        	            	: Start predicting XGBoost model...
2020-09-30T01:48:47.0551009Z [0m[0m        	            	
2020-09-30T01:48:47.0552196Z [0m[0m        	            	[09:48:27] SparsePageSource::CreateRowPage Finished writing to /tmp/kouvjl02/predict.txt_0.cache
2020-09-30T01:48:47.0552933Z [0m[0m        	            	
2020-09-30T01:48:47.0553689Z [0m[0m        	            	[09:48:27] 10x4 matrix with 30 entries loaded from /tmp/kouvjl02/predict.txt_0#/tmp/kouvjl02/predict.txt_0.cache
2020-09-30T01:48:47.0554390Z [0m[0m    executor_ir_test.go:162: 
2020-09-30T01:48:47.0554916Z [0m[0m        	Error Trace:	executor_ir_test.go:162
2020-09-30T01:48:47.0555455Z [0m[0m        	            				executor_ir_test.go:156
2020-09-30T01:48:47.0556108Z [0m[0m        	Error:      	Should be true
2020-09-30T01:48:47.0556738Z [0m[0m        	Test:       	TestExecuteXGBoostClassifier
2020-09-30T01:48:47.0557497Z [0m[0m        	Messages:   	runSQLProgram error: exit status 1
2020-09-30T01:48:47.0558581Z [0m[0m        	            	[09:48:46] SparsePage::Writer Finished writing to /tmp/7uk7rl18/predict.txt_0.cache.row.page
2020-09-30T01:48:47.0559327Z [0m[0m        	            	Traceback (most recent call last):
2020-09-30T01:48:47.0559889Z [0m[0m        	            	  File "<stdin>", line 13, in <module>
2020-09-30T01:48:47.0560849Z [0m[0m        	            	  File "/home/sqlflow/actions-runner/_work/sqlflow/sqlflow/python/runtime/local/submitter.py", line 116, in submit_local_pred
2020-09-30T01:48:47.0561621Z [0m[0m        	            	    model=model)
2020-09-30T01:48:47.0562483Z [0m[0m        	            	  File "/home/sqlflow/actions-runner/_work/sqlflow/sqlflow/python/runtime/local/xgboost_submitter/predict.py", line 93, in pred
2020-09-30T01:48:47.0563414Z [0m[0m        	            	    train_label_idx, feature_file_name, conn)
2020-09-30T01:48:47.0564398Z [0m[0m        	            	  File "/home/sqlflow/actions-runner/_work/sqlflow/sqlflow/python/runtime/local/xgboost_submitter/predict.py", line 156, in _store_predict_result
2020-09-30T01:48:47.0565201Z [0m[0m        	            	    line_no += 1
2020-09-30T01:48:47.0565807Z [0m[0m        	            	  File "/usr/lib/python3.6/contextlib.py", line 88, in __exit__
2020-09-30T01:48:47.0566393Z [0m[0m        	            	    next(self.gen)
2020-09-30T01:48:47.0567292Z [0m[0m        	            	  File "/home/sqlflow/actions-runner/_work/sqlflow/sqlflow/python/runtime/db.py", line 299, in buffered_db_writer
2020-09-30T01:48:47.0568013Z [0m[0m        	            	    w.close()
2020-09-30T01:48:47.0568798Z [0m[0m        	            	  File "/home/sqlflow/actions-runner/_work/sqlflow/sqlflow/python/runtime/db_writer/base.py", line 38, in close
2020-09-30T01:48:47.0569523Z [0m[0m        	            	    self.flush()
2020-09-30T01:48:47.0570323Z [0m[0m        	            	  File "/home/sqlflow/actions-runner/_work/sqlflow/sqlflow/python/runtime/db_writer/mysql.py", line 43, in flush
2020-09-30T01:48:47.0571297Z [0m[0m        	            	    cursor.executemany(self.statement, self.rows)
2020-09-30T01:48:47.0572404Z [0m[0m        	            	  File "/home/sqlflow/actions-runner/_work/sqlflow/sqlflow/build/env/lib/python3.6/site-packages/MySQLdb/cursors.py", line 237, in executemany
2020-09-30T01:48:47.0573338Z [0m[0m        	            	    self._get_db().encoding)
2020-09-30T01:48:47.0574306Z [0m[0m        	            	  File "/home/sqlflow/actions-runner/_work/sqlflow/sqlflow/build/env/lib/python3.6/site-packages/MySQLdb/cursors.py", line 264, in _do_execute_many
2020-09-30T01:48:47.0575195Z [0m[0m        	            	    rows += self.execute(sql + postfix)
2020-09-30T01:48:47.0576170Z [0m[0m        	            	  File "/home/sqlflow/actions-runner/_work/sqlflow/sqlflow/build/env/lib/python3.6/site-packages/MySQLdb/cursors.py", line 209, in execute
2020-09-30T01:48:47.0577012Z [0m[0m        	            	    res = self._query(query)
2020-09-30T01:48:47.0577939Z [0m[0m        	            	  File "/home/sqlflow/actions-runner/_work/sqlflow/sqlflow/build/env/lib/python3.6/site-packages/MySQLdb/cursors.py", line 315, in _query
2020-09-30T01:48:47.0578732Z [0m[0m        	            	    db.query(q)
2020-09-30T01:48:47.0579666Z [0m[0m        	            	  File "/home/sqlflow/actions-runner/_work/sqlflow/sqlflow/build/env/lib/python3.6/site-packages/MySQLdb/connections.py", line 226, in query
2020-09-30T01:48:47.0580696Z [0m[0m        	            	    _mysql.connection.query(self, query)
2020-09-30T01:48:47.0581790Z [0m[0m        	            	MySQLdb._exceptions.OperationalError: (1366, "Incorrect integer value: '[0.0063358  0.01805509 0.9756091 ]' for column `iris`.`predict`.`class` at row 1")
2020-09-30T01:48:47.0582831Z [0m[0m        	            	: Start predicting XGBoost model...
2020-09-30T01:48:47.0583320Z [0m[0m        	            	
2020-09-30T01:48:47.0584332Z [0m[0m        	            	[09:48:46] SparsePageSource::CreateRowPage Finished writing to /tmp/7uk7rl18/predict.txt_0.cache
2020-09-30T01:48:47.0585008Z [0m[0m        	            	
2020-09-30T01:48:47.0585738Z [0m[0m        	            	[09:48:46] 10x4 matrix with 30 entries loaded from /tmp/7uk7rl18/predict.txt_0#/tmp/7uk7rl18/predict.txt_0.cache
```

--- a/python/runtime/local/xgboost_submitter/predict.py
+++ b/python/runtime/local/xgboost_submitter/predict.py
@@ -115,14 +115,14 @@ def _calc_predict_result(bst, dpred, model_params):
     # not only on PAI submitter
     # TODO(yancey1989): output the original result for various
     # objective function.
-    objective = model_params.get("objective", "")
-    if objective.startswith("binary:"):
-        preds = (preds > 0.5).astype(np.int64)
-    elif objective.startswith("multi:softprob"):
-        # multi:softprob outputs probability
-        preds = np.argmax(preds, axis=1)
-    elif preds.ndim == 2:
-        preds = np.argmax(preds, axis=1)
+    obj = model_params.get("objective", "")
+    # binary:hinge output class labels
+    if obj.startswith("binary:logistic"):
+        preds = (preds > 0.5).astype(int)
+    # multi:softmax output class labels
+    elif obj.startswith("multi:softprob"):
+        preds = np.argmax(np.array(preds), axis=1)
+    # TODO(typhoonzero): deal with binary:logitraw when needed.
 
     return preds
 

--- a/python/runtime/local/xgboost_submitter/predict.py
+++ b/python/runtime/local/xgboost_submitter/predict.py
@@ -118,7 +118,8 @@ def _calc_predict_result(bst, dpred, model_params):
     objective = model_params.get("objective", "")
     if objective.startswith("binary:"):
         preds = (preds > 0.5).astype(np.int64)
-    elif objective.startswith("multi:") and preds.ndim == 2:
+    elif objective.startswith("multi:softprob"):
+        # multi:softprob outputs probability
         preds = np.argmax(preds, axis=1)
     elif preds.ndim == 2:
         preds = np.argmax(preds, axis=1)

--- a/python/runtime/local/xgboost_submitter/predict.py
+++ b/python/runtime/local/xgboost_submitter/predict.py
@@ -109,6 +109,7 @@ def _calc_predict_result(bst, dpred, model_params):
         The prediction result.
     """
     preds = bst.predict(dpred)
+    preds = np.array(preds)
 
     # TODO(yancey1989): should save train_params and model_params
     # not only on PAI submitter
@@ -117,8 +118,10 @@ def _calc_predict_result(bst, dpred, model_params):
     objective = model_params.get("objective", "")
     if objective.startswith("binary:"):
         preds = (preds > 0.5).astype(np.int64)
-    elif objective.startswith("multi:") and len(preds) == 2:
-        preds = np.argmax(np.array(preds), axis=1)
+    elif objective.startswith("multi:") and preds.ndim == 2:
+        preds = np.argmax(preds, axis=1)
+    elif preds.ndim == 2:
+        preds = np.argmax(preds, axis=1)
 
     return preds
 

--- a/python/runtime/local/xgboost_submitter/xgboost_local_test.py
+++ b/python/runtime/local/xgboost_submitter/xgboost_local_test.py
@@ -45,7 +45,7 @@ class TestXGBoostTrain(unittest.TestCase):
         original_sql = """SELECT * FROM iris.train
         TO TRAIN xgboost.gbtree
         WITH
-            objective="multi:softmax",
+            objective="multi:softprob",
             num_boost_round=20,
             num_class=3,
             validation.select="SELECT * FROM iris.test"
@@ -56,7 +56,7 @@ class TestXGBoostTrain(unittest.TestCase):
         select = "SELECT * FROM iris.train"
         val_select = "SELECT * FROM iris.test"
         train_params = {"num_boost_round": 20}
-        model_params = {"num_class": 3, "objective": "multi:softmax"}
+        model_params = {"num_class": 3, "objective": "multi:softprob"}
         save_name = "iris.xgboost_train_model_test"
         class_name = "class"
 

